futures-util/src/stream/async_reader.rs

```diff
@@ -0,0 +1,157 @@
+use core::marker::Unpin;
+use core::pin::Pin;
+use futures_core::stream::Stream;
+use futures_core::task::{LocalWaker, Poll};
+use futures_io::AsyncRead;
+use std::cmp;
+use std::io::Result;
+
+/// A stream combinator which interprets a stream of bytes as an `AsyncRead`.
+///
+/// This structure is produced by the `Stream::async_reader` method.
+#[derive(Debug)]
+#[must_use = "streams do nothing unless polled"]
+pub struct AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+    stream: St,
+    state: ReadState<St::Item>,
+}
+
+impl<St: Stream + Unpin> Unpin for AsyncReader<St>
+where
+    St: Stream + Unpin,
+    St::Item: AsRef<[u8]>,
+{
+}
+
+#[derive(Debug)]
+enum ReadState<T: AsRef<[u8]>> {
+    Ready { chunk: T, chunk_start: usize },
+    NotReady,
```

Comment:
nit: s/NotReady/Pending

Reply:
Actually, this is more `NoChunkPresent` or something, right?

Reply:
How does `PendingChunk` sound? When we are in this state, we really are in a holding pattern until the stream produces the next item.